knodle/transformation/rule_reduction.py

```diff
@@ -0,0 +1,168 @@
+    """
+    if max_rules is None and min_coverage is None:
+        logger.info("No filtering criteria ('max_lf' or 'min_coverage' for LF specified, "
+                    "returning the original LF matches.")
+        return {"rule_matches_z": rule_matches_z, "mapping_rule_class_t": mapping_rule_class_t}
+
+    coverage_per_rule = rule_matches_z.sum(0) / rule_matches_z.shape[0]
+
+    # create mask to indicate which rules will be kept unchanged
+    lf_kept_mask = np.zeros(mapping_rule_class_t.shape[0], dtype=np.bool)
+
+    # take top N rules
+    if max_rules is not None:
+        max_lf_ids = (coverage_per_rule * -1).argsort(kind="stable")[:max_rules]
+        lf_kept_mask[max_lf_ids] = True
+
+    # filter under-coveraged rules
+    if min_coverage is not None:
+        lf_kept_mask[coverage_per_rule < min_coverage] = False
+
+    # add training matches to the dictionary
+    rule_matches_dict = {"train_rule_matches_z": rule_matches_z}
+    if rule_matches_rest is not None:
+        rule_matches_dict.update(rule_matches_rest)
+
+    if drop_rules:
+        return _reduce_by_drop(rule_matches_dict, mapping_rule_class_t, lf_kept_mask)
+    else:
+        return _reduce_by_merge(rule_matches_dict, mapping_rule_class_t, lf_kept_mask)
+
+
+def _reduce_by_drop(rule_matches_dict, mapping_rule_class_t, lf_kept_mask):
+    output_dict = {}
+    for split, match_matrix in rule_matches_dict.items():
+        output_dict[split] = match_matrix[:, lf_kept_mask]
+    output_dict["mapping_rule_class_t"] = mapping_rule_class_t[lf_kept_mask, :]
+    return output_dict
+
+
+def _reduce_by_merge(rule_matches_dict, mapping_rule_class_t, lf_kept_mask):
+    # get core part of remaining rule matches
+    output_dict = _reduce_by_drop(rule_matches_dict, mapping_rule_class_t, lf_kept_mask)
+
+    # reduce all provided match matrices in the same manner; keep dict keys
+    for split, full_match_matrix in rule_matches_dict.items():
+        reduced_matches = _get_merged_matrix(
+            full_matches=full_match_matrix,
+            to_reduce_mask=~lf_kept_mask,
+            label_rule_masks=_get_lf_by_label_iterator(mapping_rule_class_t)
+        )
+
+        # add merged rules to the core matches
+        output_dict[split] = np.hstack([output_dict[split], reduced_matches])
+
+    # add merged mapping
+    merged_mapping = _get_merged_mapping(
+        to_reduce_mask=~lf_kept_mask,
+        label_rule_masks=_get_lf_by_label_iterator(mapping_rule_class_t),
+        number_of_labels=mapping_rule_class_t.shape[1]
+    )
+    output_dict["mapping_rule_class_t"] = np.vstack([output_dict["mapping_rule_class_t"], merged_mapping])
+
+    return output_dict
+
+
+def _get_lf_by_label_iterator(
+    mapping_rule_class_t: np.ndarray
+) -> Iterable[np.ndarray]:
+    """
+    Get an iterator yielding an rule ids mask for every label, indicating which rules belong to this label.
```

Comment:
an -> a

Reply:
btw there is a function `get_rules_sample_ids` in knodle/trainer/wscrossweigh - I guess it is done pretty much the same, but not exactly.